src/exp/istio-watcher/pkg/watcher/watcher.go

```diff
@@ -0,0 +1,291 @@
+		return err
+	}
+
+	for _, pod := range podList.Items {
+		// Known for loop gotcha with goroutines
+		curr := pod
+		sendersErrGroup.Go(func() error {
+			if err := m.getEnvoyMetricsFromSidecar(sendersCtx, curr, metricsChan); err != nil {
+				logrus.Errorf("Failed fetching request metrics from pod %s", curr.Name)
+				return err
+			}
+			return nil
+		})
+	}
+	receiversErrGroup.Go(func() error {
+		// Function call below updates a map which isn't concurrent-safe.
+		// Needs to be taken into consideration if the code should ever change to use multiple goroutines
+		if err := m.convertMetricsToConnections(metricsChan, done); err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err := sendersErrGroup.Wait(); err != nil {
+		return err
+	}
+	done <- 0
+
+	if err := receiversErrGroup.Wait(); err != nil {
+		return err
+	}
+
+	close(metricsChan)
+	return nil
+}
+
+func (m *IstioWatcher) getEnvoyMetricsFromSidecar(ctx context.Context, pod corev1.Pod, metricsChan chan<- *EnvoyMetrics) error {
+	req := m.clientset.CoreV1().
+		RESTClient().
+		Post().
+		Resource("pods").
+		Name(pod.Name).
+		Namespace(pod.Namespace).SubResource("exec").
+		VersionedParams(&corev1.PodExecOptions{
+			Command:   strings.Split(IstioProxyTotalRequestsCMD, " "),
+			Stdout:    true, // We omit stderr and we error according to return code from executed cmd
+			Container: IstioSidecarContainerName,
+		}, scheme.ParameterCodec)
+
+	// TODO: use error group context in exec
+	exec, err := remotecommand.NewSPDYExecutor(m.config, "POST", req.URL())
+	if err != nil {
+		return err
+	}
+
+	var outBuf bytes.Buffer
+	streamOpts := remotecommand.StreamOptions{Stdout: &outBuf}
+	err = exec.StreamWithContext(ctx, streamOpts)
+	if err != nil {
+		return err
+	}
+
+	metrics := &EnvoyMetrics{}
+	if err := json.NewDecoder(&outBuf).Decode(metrics); err != nil {
+		return err
+	}
+	metricsWithPath := make([]Metric, 0)
+	for _, metric := range metrics.Stats {
+		if strings.Contains(metric.Name, "request_path") {
+			metricsWithPath = append(metricsWithPath, metric)
+		}
+	}
+
+	metrics.Stats = metricsWithPath
+	if len(metrics.Stats) == 0 {
+		return nil
+	}
+
+	metricsChan <- metrics
+	return nil
+}
+
+func (m *IstioWatcher) convertMetricsToConnections(metricsChan <-chan *EnvoyMetrics, done <-chan int) error {
+	for {
+		select {
+		case metrics := <-metricsChan:
+			for _, metric := range metrics.Stats {
+				conn, err := m.buildConnectionFromMetric(metric)
+				if err != nil && errors.Is(err, ConnectionInfoInSufficient) {
+					continue
+				}
+				if err != nil {
+					return err
+				}
+				m.connections[conn] = time.Now()
+			}
+		case <-done:
+			logrus.Debugf("Got done signal")
+			return nil
+		}
+	}
+}
+
+func (m *IstioWatcher) buildConnectionFromMetric(metric Metric) (*ConnectionWithPath, error) {
+	conn := &ConnectionWithPath{}
+	err := EnvoyConnectionMetricRegex.MatchToTarget(metric.Name, conn)
+	if err != nil && errors.Is(err, &regroup.NoMatchFoundError{}) {
```

Comment:
Does this work? Doesn't `Is` need to work with a specific instance? (as in a sentinel error)